Sources/ClientRuntime/Networking/Http/URI.swift

```diff
@@ -0,0 +1,148 @@
+    }
+
+    public func toBuilder() -> URIBuilder {
+        return URIBuilder()
+           .withScheme(self.scheme)
+           .withPath(self.path)
+           .withHost(self.host)
+           .withPort(self.port)
+           .withQueryItems(self.query)
+    }
+}
+
+public class URIBuilder {
+    var urlComponents: URLComponents
+    var scheme: Scheme = Scheme.https
+    var path: String = "/"
+    var host: String = ""
+    var port: Int16 = Int16(Scheme.https.port)
+    var query: [SDKURLQueryItem] = []
+    var username: String?
+    var password: String?
+
+    required public init() {
+        self.urlComponents = URLComponents()
+        self.urlComponents.scheme = self.scheme.rawValue
+        self.urlComponents.path = self.path
+        self.urlComponents.host = self.host
+    }
+
+    @discardableResult
+    public func withScheme(_ value: Scheme?) -> URIBuilder {
+        self.scheme = value ?? Scheme.https
+        self.urlComponents.scheme = self.scheme.rawValue
+        return self
+    }
+
+    @discardableResult
+    public func withPath(_ value: String) -> URIBuilder {
+        self.path = value
+        if self.path.contains("%") {
+            self.urlComponents.percentEncodedPath = self.path
+        } else {
+            self.urlComponents.path = self.path
+        }
+        return self
+    }
+
+    @discardableResult
+    public func withHost(_ value: String) -> URIBuilder {
+        self.host = value
+        self.urlComponents.host = self.host
+        return self
+    }
+
+    @discardableResult
+    public func withPort(_ value: Int16) -> URIBuilder {
+        self.port = value
+        return self
+    }
+
+    @discardableResult
+    public func withQueryItems(_ value: [SDKURLQueryItem]) -> URIBuilder {
+        self.query.append(contentsOf: value)
+        if !self.query.isEmpty {
+            self.urlComponents.percentEncodedQuery = self.query.map { queryItem in
+                            return [queryItem.name, queryItem.value].compactMap { $0 }.joined(separator: "=")
+                        }.joined(separator: "&")
+        }
+        return self
+    }
+
+    @discardableResult
+    public func withQueryItem(_ value: SDKURLQueryItem) -> URIBuilder {
+        withQueryItems([value])
+    }
+
+    @discardableResult
+    public func withUsername(_ value: String) -> URIBuilder {
+        self.username = value
+        self.urlComponents.user = self.username
+        return self
+    }
+
+    @discardableResult
+    public func withPassword(_ value: String) -> URIBuilder {
+        self.password = value
+        self.urlComponents.password = self.password
+        return self
+    }
+
+    public func build() -> URI {
+        return URI(scheme: self.scheme,
+                   path: self.path,
+                   host: self.host,
+                   port: self.port,
+                   query: self.query,
+                   username: self.username,
+                   password: self.password)
```

Comment:
If my understanding is correct, `URIBuilder` is using `URLComponents` to throw errors when an invalid value is passed. But `URLComponents` also percent-encodes stuff when necessary - like if you did `urlComponents.query = "#"`, it'll encode the query to be `"%23"`. I think you could just use `URLComponents` instead of storing the other members.